### PR TITLE
add Support for CycloneV SoC ethernet

### PR DESCRIPTION
**Pull request by mevalver**
drivers: ethernet: add Support for CycloneV SoC ethernet
Adding driver for ethernet. Cyclone V SoC-specific driver and config
files

Signed-off-by: Esteban Valverde <esteban.valverde.vega@intel.com>

### CODE REVIEW

**Review by npitre — COMMENTED** [review]
Some comments.

Please don't reuse `dwmac_*` symbol and filename prefixes. Either this is 
similar to the existing dwmac driver in which case you should be reusing 
the eth_dwmac.c code with a platform-specific shim on top of it like 
eth_dwmac_stm32h7x.c, or this is different enough to warrant a separate 
driver entirely (e.g. eth_cyclonev.c, cyclonev_send(), etc.)

Then you should really consider avoiding data payload copy to get more 
performance out of your driver. That memcpy() in your dwmac_send() and
net_pkt_write() in your dwmac_receive() could be skipped if you implement 
a zero-copy strategy. That's what the ethh_dwmac.c driver does which 
appears to be the inspiration for your driver. The help text for 
CONFIG_DWMAC_NB_TX_DESCS and CONFIG_DWMAC_NB_RX_DESCS copied from the 
eth_dwmac.c driver was redacted with zero-copy operation in mind and 
therefore aren't fully appropriate for your driver's current state.

File: dts/bindings/ethernet/snps,ethernet-cyclonev.yaml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright (C) 2022, Intel Corporation
+# Description:
+# Binding file for Ethernet driver CycloneV
+# heavily modified for Zephyr
+
+description: Synopsys DesignWare Ethernet
+
+compatible: "snps,ethernet-cyclonev"
```
Comment (galak) [inline dts/bindings/ethernet/snps,ethernet-cyclonev.yaml:9]: This compatible doesn't feel right.  Is there something cyclonev specific about the ethernet device?  Is this not some `snps,dwmac-<IP_VER>`?

https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/net/snps,dwmac.yaml

**Review by galak — CHANGES_REQUESTED** [review]
Feels like this should be a bit more generic from a devicetree and naming convention if its supporting some snps,ethernet IP.

**Comment by mevalver** [comment]
Regarding your comments

We of course considered reusing any existing driver that was already in Zephyr. Either the dwmac. However, after doing our analysis and trying to make it work with our Cyclone V SoC hardware, my team concluded that the existing dwmac driver or other variances were not satisfying our needs. So we decided to write our own, based on the knowledge of the SoC.

For this reason, I have changed the "dwmac" symbols with "eth_cyclone" in all my files, making sure there is no confusion.

As per your second observation. For now, we are not interested in the best performance.

File: dts/bindings/ethernet/snps,ethernet-cyclonev.yaml (hunk 1 above)
Comment (mevalver) [reply dts/bindings/ethernet/snps,ethernet-cyclonev.yaml:9]: Actually the one we use for linux is this one:
https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/net/socfpga-dwmac.txt

I can change the naming, but this is up to you...

Comment (mevalver) [reply dts/bindings/ethernet/snps,ethernet-cyclonev.yaml:9]: Which is compatible with CycloneV, Arria10, Stratix10... But I dont want to use "socfpga" as some other FPGAs might not work with it (Like Agilex)

**Comment by mevalver** [comment]
Could you please specify the actions to make this available?

**Comment by mevalver** [comment]
Hi @galak  @npitre ,
I was consulting one of our more experienced embedded software engineers about this driver.
In a nutshell, he points out the following:

- At this point moving the driver to a "zero-copy" strategy would mean us to do an additional effort. We are looking for a driver that is simple and reliable, this is something that we have achieved with this PR.

-  There are other ethernet drivers in Zephyr that also use buffers for sending. E.g Xilinx's GEM driver, function eth_xlnx_gem_send: 

`/* Copy packet data to DMA buffer */
                             net_pkt_read(pkt, (void *)tx_buffer_offs,
                                                (tx_data_remaining < dev_conf->tx_buffer_size) ?
                                                tx_data_remaining : dev_conf->tx_buffer_size);
`
- The net_pkt_read calls memcpy via net_pkt_cursor_operate in subsys/net/ip/net_pkt
-  The advantage of the memcpy is that the buffer can be reused after the call to eth_cyclonev_send,  otherwise the application has to track when each packet is sent which is more effort for anyone using the driver.

Could you please advise what we have to change to accept this driver?

Thanks,
E.V.

**Comment by npitre** [comment]
Going zero-copy was a suggestion not a blocker on my part.

**Comment by galak** [comment]
@mevalver the point was not about zero-copy or not.  Its about what is the hardware, is there a driver already for that hardware (or something very similar) and not duplicating or creating new variants (of both driver and dts bindings).

File: dts/bindings/ethernet/snps,ethernet-cyclonev.yaml (hunk 1 above)
Comment (mevalver) [reply dts/bindings/ethernet/snps,ethernet-cyclonev.yaml:9]: @galak just to be 100% clear, should I go ahead an change the naming to "altr,socfpga-stmmac" everywhere?

**Pull request closed without merging**